pennylane/beta/queuing/queuing.py

```diff
@@ -0,0 +1,188 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
+import abc
+from collections import OrderedDict, deque
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for objects.
+
+    In PennyLane, the construction of quantum gates is separated from the specific
+    quantum node (:class:`~.BaseQNode`) that they belong to. However, including
+    logic for this when creating an instance of :class:`~.Operator` does not align
+    with the current architecture. Therefore, there is a need to use a high level
+    object that holds information about the relationship between quantum gates and
+    a quantum node.
+
+    The :class:`~.QueuingContext` class realizes this by providing access to
+    the current QNode.  Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
+
+    The ``QueuingContext`` class both acts as the abstract base class for all
+    classes that expose a queue for Operations (so-called contexts), as well as the
+    interface to said queues. The active contexts contain maximally one QNode and
+    an arbitrary number of other contexts like the :class:`~.OperationRecorder`.
+    """
+
+    # TODO: update docstring
+
+    _active_contexts = deque()
+    """The stack of contexts that are currently active."""
+
+    def __enter__(self):
+        """Adds this instance to the global list of active contexts.
+
+        Returns:
+            QueuingContext: This instance
+        """
+        QueuingContext._active_contexts.append(self)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Remove this instance from the global list of active contexts.
+        """
+        QueuingContext._active_contexts.remove(self)
+
+    @abc.abstractmethod
+    def _append(self, obj, **kwargs):
+        """Append an object to this QueuingContext instance.
+
+        Args:
+            obj: The object to be appended
+        """
+
+    @classmethod
+    def active_context(cls):
+        """Returns the currently active queuing context."""
+        if cls._active_contexts:
+            return cls._active_contexts[-1]
+
+        return None
+
+    @classmethod
+    def append(cls, obj, **kwargs):
+        """Append an object to the queue(s).
+
+        Args:
+            obj: the object to be appended
+        """
+
+        # Beta:
+        # For now, only append if there exists an active_context
+        active_context = cls.active_context()
+
+        if active_context:
+            active_context._append(obj, **kwargs)  # pylint: disable=protected-access

```

Comment:
Added this logic such to allow examples where there's no active context when an operation is created, but it is added to the queue in a supported way, such as
```python
            CNOT = qml.CNOT(wires=[0, 1])

            def circuit(x):
                qml.QueuingContext.append(CNOT)
                qml.RY(0.4, wires=[0])
                qml.RZ(-0.2, wires=[1])

                return qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(1))
```